File: src/python/clear_text_logging.py
"""CWE-532 / CWE-312: Clear-text logging / storage of sensitive data.

Rules: py/clear-text-logging-sensitive-data, py/clear-text-storage-sensitive-data
Autofix: typically NOT provided — needs PII / credential classification and
a redaction strategy.
"""

import logging
import json

logging.basicConfig(level=logging.INFO)
log = logging.getLogger("auth")


def login(username: str, password: str, card_number: str | None = None) -> None:
    log.info("login attempt user=%s password=%s", username, password)
    if card_number:
        log.warning("payment card=%s", card_number)
    with open("/var/log/app/audit.log", "a") as f:
        f.write(json.dumps({"user": username, "password": password}) + "\n")


def store_user(username: str, password: str) -> None:
    with open("/var/app/users.txt", "a") as f:
        f.write(f"{username}:{password}\n")

File: src/javascript/clear-text-logging.js
// CWE-312 / CWE-532: Clear-text logging / storage of sensitive data.
// Rules: js/clear-text-logging, js/clear-text-storage-of-sensitive-information
// Autofix: typically NOT provided — fix requires deciding which fields are
// sensitive (PII vs. credentials), introducing a redaction layer, and
// potentially changing the log transport.

const express = require('express');
const fs = require('fs');

const app = express();
app.use(express.json());

app.post('/login', (req, res) => {
  console.log('Login attempt', JSON.stringify(req.body));
  fs.appendFileSync('/var/log/app/auth.log',
    `${new Date().toISOString()} user=${req.body.username} pass=${req.body.password}\n`);
  res.json({ ok: true });
});

app.post('/payment', (req, res) => {
  console.info('Charging card', req.body.cardNumber, 'cvv', req.body.cvv);
  res.json({ ok: true });
});

app.listen(3111);
